libgadget/fof.c

```diff
@@ -442,33 +424,38 @@ fofp_merge(int target, int other, TreeWalk * tw)
 {
     /* this will lock h1 */
     int * Head = FOF_PRIMARY_GET_PRIV(tw)->Head;
-    struct SpinLocks * spin = FOF_PRIMARY_GET_PRIV(tw)->spin;
     int h1, h2;
-
     do {
-        h1 = HEADl(-1, target, -1, Head, spin);
-        /* stop looking if we find h1 along the path (because it is already owned by us) */
-        h2 = HEADl(h1, other, h1, Head, spin);
-        /* We had a lock already taken on h2 by another thread.
-         * We need to unlock h1 and retry to avoid deadlock loops.*/
-        if(h2 == -2)
-            unlock_spinlock(h1, spin);
-    } while(h2 == -2);
-
-    if(h2 >=0)
-    {
-        /* h2 as a sub-tree of h1 */
-        Head[h2] = h1;
-
-        /* update MinID of h1 */
-        if(HaloLabel[h1].MinID > HaloLabel[h2].MinID)
-        {
-            HaloLabel[h1].MinID = HaloLabel[h2].MinID;
-            HaloLabel[h1].MinIDTask = HaloLabel[h2].MinIDTask;
+        h1 = HEADl(-1, target, Head);
+        /* Done if we find h1 along the path
+         * (because other is already in the same halo) */
+        h2 = HEADl(h1, other, Head);
+        if(h2 < 0)
+            return;
+        /* Ensure that we always merge to the lower entry*/
+        if(h1 > h2) {
```

Comment:
I somehow thought we tried this before an run into a bug. Do you recall an event like that?

Reply:
Yes, we tried it before and hit a deadlocking bug: we were taking mutexes on both h1 and h2 and had to be careful of the order. Because this is lockless we don't have that problem anymore. There is still a circular reference problem, which is why the swap. I will add a comment.